src/crypto/public_key/rsa.js

```diff
@@ -18,254 +18,205 @@
+    if (n.cmp(m) <= 0) {
+      throw new Error('Data too large.');
+    }
+    const dq = d.mod(q.subn(1)); // d mod (q-1)
+    const dp = d.mod(p.subn(1)); // d mod (p-1)
+    const pred = new BN.red(p);
+    const qred = new BN.red(q);
+    const nred = new BN.red(n);
+
+    let blinder;
+    let unblinder;
+    if (config.rsa_blinding) {
+      unblinder = random.getRandomBN(two, n).toRed(nred);
+      blinder = unblinder.redInvm().redPow(e);
+      m = m.toRed(nred).redMul(blinder).fromRed();
+    }
 
-  function verify(x, e, n) {
-    return x.modPowInt(e, n);
-  }
+    const mp = m.toRed(pred).redPow(dp);
+    const mq = m.toRed(qred).redPow(dq);
+    const t = mq.redSub(mp.fromRed().toRed(qred));
+    const h = u.toRed(qred).redMul(t).fromRed();
 
-  // "empty" RSA key constructor
+    let result = h.mul(p).add(mp).toRed(nred);
 
-  function KeyObject() {
-    this.n = null;
-    this.e = 0;
-    this.ee = null;
-    this.d = null;
-    this.p = null;
-    this.q = null;
-    this.dmp1 = null;
-    this.dmq1 = null;
-    this.u = null;
-  }
+    if (config.rsa_blinding) {
+      result = result.redMul(unblinder);
+    }
 
-  // Generate a new random private key B bits long, using public expt E
+    return result.toArrayLike(Uint8Array, 'be', n.byteLength());
+  },
 
-  function generate(B, E) {
+  /**
+   * Generate a new random private key B bits long with public exponent E
+   * @param {Integer} B RSA bit length
+   * @param {String}  E RSA public exponent in hex string
+   * @return {{n: BN, e: BN, d: BN,
+               p: BN, q: BN, u: BN}} RSA public modulus, RSA public exponent, RSA private exponent,
+                                     RSA private prime p, RSA private prime q, u = q ** -1 mod p
+   */
+  generate: async function(B, E) {
+    let key;
+    E = new BN(E, 16);
     const webCrypto = util.getWebCryptoAll();
 
-    //
     // Native RSA keygen using Web Crypto
-    //
-
     if (webCrypto) {
-      const Euint32 = new Uint32Array([parseInt(E, 16)]); // get integer of exponent
-      const Euint8 = new Uint8Array(Euint32.buffer); // get bytes of exponent
+      let keyPair;
       let keyGenOpt;
-
-      let keys;
       if ((window.crypto && window.crypto.subtle) || window.msCrypto) {
         // current standard spec
         keyGenOpt = {
           name: 'RSASSA-PKCS1-v1_5',
           modulusLength: B, // the specified keysize in bits
-          publicExponent: Euint8.subarray(0, 3), // take three bytes (max 65537)
+          publicExponent: E.toArrayLike(Uint8Array), // take three bytes (max 65537) for exponent
           hash: {
             name: 'SHA-1' // not required for actual RSA keys, but for crypto api 'sign' and 'verify'
           }
         };
-
-        keys = webCrypto.generateKey(keyGenOpt, true, ['sign', 'verify']);
-        if (typeof keys.then !== 'function') { // IE11 KeyOperation
-          keys = util.promisifyIE11Op(keys, 'Error generating RSA key pair.');
-        }
```

Comment:
This is not needed?

Reply:
Which part? The `promisifyIE11Op` is not needed because `await` can be used for both promises or regular return values.

Reply:
Yeah, but in IE's case the return value is not a regular return value, it's a KeyOperation object which is it's own async thing. I would think you would still need this...

Reply:
Ah, sorry I missed that. I'll revert this change in a new commit.